src/restart_windows.nim

```diff
@@ -0,0 +1,86 @@
+## Procedures necessary for implementing :restart on Windows
+
+import winlean, os
+
+template doWhile(a: bool, b:untyped): untyped =
+    b
+    while a:
+        b
+
+type
+    ## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/ns-tlhelp32-processentry32
+    PROCESSENTRY32 = object
+        dwSize: DWORD
+        cntUsage: DWORD
+        th32ProcessID: DWORD
+        th32DefaultHeapID: ULONG_PTR
+        th32ModuleID: DWORD
+        cntThreads: DWORD
+        th32ParentProcessID: DWORD
+        pcPriClassBase: LONG
+        dwFlags: DWORD
+        # winlean does not include the CHAR type, but Windows.h just typedefs CHAR to char anyway.
+        # Win32 API type names are otherwise used for consistency with the documentation.
+        szExeFile: array[MAX_PATH, cchar]
+
+## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot
+proc createToolhelp32Snapshot(dwFlags: DWORD, th32ProcessID: DWORD): Handle
+    {.stdcall, dynlib: "kernel32", importc: "CreateToolhelp32Snapshot".}
+
+## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-process32first
+proc process32First(hSnapshot: Handle, lppe: var PROCESSENTRY32): bool
+    {.stdcall, dynlib: "kernel32", importc: "Process32First".}
+
+## See https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-process32next
+proc process32Next(hSnapshot: Handle, lppe: var PROCESSENTRY32): bool
+    {.stdcall, dynlib: "kernel32", importc: "Process32Next".}
+
+## Unlike POSIX getppid, can be passed an arbitrary PID.
+proc getppid(pid = getCurrentProcessId().DWORD): DWORD =
+    # dwFlags=2 causes the snapshot to include all currently running processes
+    # th32ProcessID=0 refers to the calling process
+    let handle = createToolhelp32Snapshot(dwFlags = 2, th32ProcessID = 0)
+    var processEntry: PROCESSENTRY32
+    processEntry.dwSize = sizeof(PROCESSENTRY32).DWORD
+
+    if process32First(handle, processEntry):
```

Comment:
if `process32First` can ever be false, surely this should throw an error from an `else:` branch?

Reply:
We won't be able to do much about it if this does fail, but we can tell our users we know something went wrong. Will add.

Reply:
`createToolhelp32Snapshot` or `process32Next` might also fail, so we can just raise an exception if we ever reach the end of `getppid`. If we haven't returned before getting to that point, we know something must have failed.